README.md
the_eye


changelog.txt
Build: 1.15.2-31.1.48 - Thu Apr 23 12:57:42 GMT 2020
	David Quintana:
		Add tickRate setter to the ForgeFlowingFluid Properties builder.

=========
Build: 1.15.2-31.1.47 - Mon Apr 20 01:37:16 GMT 2020
	cpw:
		Fix recursion mess in PlayerAdvancement loading, by using an alternative based on a toposorted list of all advancements.
		It seems to be fully equivalent in testing, but there may be corner cases, so there is a config to disable, though disabling
		may break servers, especially those chunkloading fake players.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.46 - Wed Apr 15 17:10:05 GMT 2020
	lex:
		Revert defualt jvm args.

=========
Build: 1.15.2-31.1.45 - Wed Apr 15 06:54:47 GMT 2020
	lex:
		Fixed dev time compilation using J9+ producing incorrect bytecode.

=========
Build: 1.15.2-31.1.44 - Sun Apr 12 19:27:19 GMT 2020
	David Quintana:
		Add "type" field to global loot modifier jsons. (#6554)
		
		* Add "type" field to global loot modifier jsons and fix tests.
		To avoid binary breaking, "type" defaults to be the same as the filename.
		* Fix license headers.

=========
Build: 1.15.2-31.1.43 - Sat Apr 11 06:03:58 GMT 2020
	tterrag:
		Fix child annotations not inheriting member name, cleanup

=========
Build: 1.15.2-31.1.42 - Sat Apr 11 05:20:15 GMT 2020
	tterrag:
		Fix element type of scanned child annotations

=========
Build: 1.15.2-31.1.41 - Thu Apr 09 00:19:35 GMT 2020
	28677702+patrick-aoc:
		Fixed PlantType not extending IExtensibleEnum, Closes #6286

=========
Build: 1.15.2-31.1.40 - Thu Apr 09 00:12:15 GMT 2020
	46117728+valoeghese:
		Added hook to allow modded Hill Biomes (#6571)

=========
Build: 1.15.2-31.1.39 - Wed Apr 08 01:45:04 GMT 2020
	tterrag:
		Revive forge lighting pipeline, disabled by default for now
		
		Remove a lot of light value convolution by using 0..1 for nearly everything
		Fix a lot of TODOs that are needed for the lighting pipeline to function
		Potential fix for #6425

	tterrag:
		Improve performance of VertexBufferConsumer by inlining element locs

=========
Build: 1.15.2-31.1.37 - Sun Apr 05 04:50:22 GMT 2020
	tterrag:
		Fix Tag serializing empty optionals, improve RegistryObject error msg

=========
Build: 1.15.2-31.1.36 - Sat Apr 04 19:36:37 GMT 2020
	tterrag:
		Remove SideOnly from WoodType methods

=========
Build: 1.15.2-31.1.35 - Wed Apr 01 19:38:28 GMT 2020
	lex:
		Cleanup IForgeTagBuilder code style.

=========
Build: 1.15.2-31.1.34 - Wed Apr 01 00:16:26 GMT 2020
	tterrag:
		Add support for optional tag values to Tag.Builder
		
		Add optional tag values to data gen test
		Fix data gen test not running by correcting gui_light value

	tterrag:
		Move global loot test data files out of generated folder

=========
Build: 1.15.2-31.1.32 - Tue Mar 31 23:57:39 GMT 2020
	tterrag:
		Deprecate ModelBuilder#gui3d, add ModelBuilder#guiLight

	tterrag:
		Fix a few issues with BufferBuilder
		
		Fix vanilla bug that caused the byte order of the buffer from getNextBuffer to be incorrect
		Fix missing modification to nextElementBytes in putBulkData

=========
Build: 1.15.2-31.1.30 - Fri Mar 27 23:26:45 GMT 2020
	naqaden:
		Add renderer to RenderNameplateEvent (#6562)

=========
Build: 1.15.2-31.1.29 - Fri Mar 27 23:19:42 GMT 2020
	dags-:
		Defer writing the server.properties file until after mods have loaded to prevent config loss (#6545)

=========
Build: 1.15.2-31.1.28 - Fri Mar 27 23:10:32 GMT 2020
	36975818+hnOsmium0001:
		Fix ownership leak in ItemStackHandler (#6580)
		
		- Triggered by invoking ItemStackHandler#extractItem with `simulate=true`

=========
Build: 1.15.2-31.1.27 - Sat Mar 21 00:06:03 GMT 2020
	cpw:
		Fix cycle happening in noteblock even though we're setting the new value as well..
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.26 - Fri Mar 20 21:35:06 GMT 2020
	cpw:
		Add a progress message for atlas textures. Fills a big time gap on bigger packs.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.25 - Tue Mar 17 09:41:49 GMT 2020
	bjfccb:
		Remove ArrowEntity patch, vanilla fixed MC-107941

=========
Build: 1.15.2-31.1.24 - Sun Mar 15 23:28:36 GMT 2020
	lex:
		Fix Forge's tps and gen commands.

=========
Build: 1.15.2-31.1.23 - Sun Mar 15 20:55:21 GMT 2020
	lex:
		Update gradle wrapper
		Update MCPConfig to version that bypasses javac compiler bugs
		Fix Global loot test compile errors.

=========
Build: 1.15.2-31.1.22 - Sun Mar 15 07:39:18 GMT 2020
	tterrag:
		Fix loading text not rendering after mojang logo appears

=========
Build: 1.15.2-31.1.21 - Sun Mar 15 07:28:02 GMT 2020
	tterrag:
		Fix loading GUI corrupting matrix state

=========
Build: 1.15.2-31.1.20 - Sun Mar 15 03:29:03 GMT 2020
	cpw:
		Somewhat restore the loading screen overlay. It seems that once the mojang logo and progress bar
		start running, I can no longer write to the window, even though I am obviously trying to..
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.19 - Wed Mar 11 07:00:30 GMT 2020
	17338378+Nightenom:
		Add projection matrix to RenderWorldLastEvent (#6536)

=========
Build: 1.15.2-31.1.18 - Thu Feb 27 06:13:47 GMT 2020
	draco18s:
		Global Loot Functions (#6401)

=========
Build: 1.15.2-31.1.17 - Wed Feb 26 03:45:24 GMT 2020
	Colby Prince:
		Updated ForgeDev MCP Mapping Version (#6532)

=========
Build: 1.15.2-31.1.16 - Mon Feb 24 03:17:00 GMT 2020
	cpw:
		Fix up alarming security crisis with network handling that allowed for wrong code execution on the server, resulting in CRASHED SERVERS. AWFUL stuff.
		
		Also fixed a silly log message from the login handler.
		
		This change introduces a mechanism to direct certain packets to only process on one side or another. Invalid sidedness will result in the connection being terminated.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.15 - Sun Feb 23 17:32:42 GMT 2020
	julian.dunskus:

src/main/java/com/seirius/theeye/TheEye.java
package com.seirius.theeye;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.io.ByteStreams;
import com.seirius.theeye.common.TheMap;
import com.sun.net.httpserver.HttpServer;
import net.minecraft.block.Block;
import net.minecraft.world.dimension.DimensionType;
import net.minecraft.world.server.ServerWorld;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.InterModComms;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.event.lifecycle.InterModEnqueueEvent;
import net.minecraftforge.fml.event.lifecycle.InterModProcessEvent;
import net.minecraftforge.fml.event.server.FMLServerStartingEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

@Mod("theeye")
public class TheEye {
    private static final Logger LOGGER = LogManager.getLogger();

    private final static LoadingCache<String, byte[]> IMAGE_MAP_CACHE = CacheBuilder.newBuilder()
            .maximumSize(3000)
            .expireAfterAccess(12, TimeUnit.HOURS)
            .build(
                    new CacheLoader<String, byte[]>() {
                        @Override
                        public byte[] load(String key) throws Exception {
                            int[] data = Arrays.stream(key.split(":")).mapToInt(Integer::parseInt).toArray();
                            return TheMap.getChunkImageAsBytes(WORLD, data[0], data[1], data[2]);
                        }
                    }
            );

    public static ServerWorld WORLD;

    public TheEye() {
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::setup);
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::enqueueIMC);
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::processIMC);
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::doClientStuff);

        MinecraftForge.EVENT_BUS.register(this);
    }

    private void setup(final FMLCommonSetupEvent event) {
    }

    private void doClientStuff(final FMLClientSetupEvent event) {
        LOGGER.info("Got game settings {}", event.getMinecraftSupplier().get().gameSettings);
    }

    private void enqueueIMC(final InterModEnqueueEvent event) {
        InterModComms.sendTo("theeye", "helloworld", () -> {
            LOGGER.info("Hello from the SeiRiuS");
            return "The eye is watching them";
        });
    }

    private void processIMC(final InterModProcessEvent event) {
    }

    @SubscribeEvent
    public void onServerStarting(FMLServerStartingEvent event) {
        WORLD = event.getServer().getWorld(DimensionType.OVERWORLD);
        System.out.println("--------------- ServerSTARTUP");
        TheMap.register(event.getCommandDispatcher());
        try {
            final String apiMap = "/api/map/";
            final String map = "/";
            HttpServer server = HttpServer.create(new InetSocketAddress(9000), 0);
            server.createContext(map, (httpExchange -> {
                try {
                    byte[] indexHtml = ByteStreams.toByteArray(TheEye.class.getResourceAsStream("/index.html"));
                    ClassLoader.getSystemClassLoader().getResource("index.html");
                    httpExchange.getResponseHeaders().set("Content-Type", "text/html");
                    httpExchange.sendResponseHeaders(200, indexHtml.length);
                    OutputStream output = httpExchange.getResponseBody();
                    output.write(indexHtml);
                    output.close();
                    httpExchange.close();
                } catch (Exception e) {
                    if (e instanceof IOException) {
                        return;
                    }
                    e.printStackTrace();
                }
            }));
            server.createContext(apiMap, (httpExchange -> {
                try {

                    String path = httpExchange.getRequestURI().getPath();
                    String params = path.substring(apiMap.length()).replace(".png", "");
                    String[] zoomXZ = params.split("/");
                    int zoom = Integer.parseInt(zoomXZ[0]);
                    int x = Integer.parseInt(zoomXZ[1]);
                    int z = Integer.parseInt(zoomXZ[2]);
                    byte[] image = IMAGE_MAP_CACHE.get(getKey(x, z, TheMap.transformZoom(zoom)));
                    httpExchange.getResponseHeaders().set("Content-Type", "image/png");
                    httpExchange.sendResponseHeaders(200, image.length);
                    OutputStream output = httpExchange.getResponseBody();
                    output.write(image);
                    output.close();
                    httpExchange.close();
                } catch (Exception e) {
                    if (e instanceof IOException) {
                        return;
                    }
                    e.printStackTrace();
                }
            }));
            server.setExecutor(null);
            server.start();
            System.out.println("Server started");
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD)
    public static class RegistryEvents {
        @SubscribeEvent
        public static void onBlocksRegistry(final RegistryEvent.Register<Block> blockRegistryEvent) {
        }
    }

    public static String getKey(int x, int z, int zoom) {
        return String.format("%d:%d:%d", x, z, zoom);
    }

}


src/main/java/com/seirius/theeye/common/TheMap.java
package com.seirius.theeye.common;

import com.mojang.brigadier.CommandDispatcher;
import net.minecraft.block.BlockState;
import net.minecraft.command.CommandSource;
import net.minecraft.command.Commands;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.chunk.IChunk;
import net.minecraft.world.gen.Heightmap;
import net.minecraft.world.server.ServerWorld;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class TheMap {

    public static int CHUNK_SIZE = 16;

    public static int MAX_ZOOM = 12;

    public static int transformZoom(int incommingZoom) {
        return MAX_ZOOM - incommingZoom + 1;
    }

    public static void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
            Commands.literal("testmap").requires((permission) -> permission.hasPermissionLevel(4))
                .then(Commands.literal("tiles").executes((command) -> {
                    return 1;
                }))
        );
    }

    public static BufferedImage getChunkImage(ServerWorld world, int x, int z, int zoom) {
        int zoomedSize = CHUNK_SIZE * zoom;
        int[] pixels = new int[zoomedSize * zoomedSize];

        int worldStartZ = z * zoomedSize;
        int worldStartX = x * zoomedSize;

        int minZ = z * zoomedSize;
        int minX = x * zoomedSize;
        int currentChunkX = resolveCoorForChunk(minX) / CHUNK_SIZE;
        int currentChunkZ = resolveCoorForChunk(minZ) / CHUNK_SIZE;

        ChunkStorage chunkStorage = new ChunkStorage(world);

        IChunk chunk = chunkStorage.getChunk(minX, minZ);

        for (int imageZ = 0; imageZ < zoomedSize; imageZ++) {
            int worldZ = imageZ + worldStartZ;
            for (int imageX = 0; imageX < zoomedSize; imageX++) {
                int worldX = imageX + worldStartX;

                int newCurrentX = resolveCoorForChunk(worldX) / CHUNK_SIZE;
                int newCurrentZ = resolveCoorForChunk(worldZ) / CHUNK_SIZE;
                if (newCurrentX != currentChunkX || newCurrentZ != currentChunkZ) {
                    currentChunkX = newCurrentX;
                    currentChunkZ = newCurrentZ;
                    chunk = chunkStorage.getChunk(worldX, worldZ);
                }
                int topY = chunk.getTopBlockY(Heightmap.Type.WORLD_SURFACE, worldX, worldZ);
                BlockState blockState = chunk.getBlockState(new BlockPos(worldX, topY, worldZ));
                pixels[imageZ * zoomedSize + imageX] = blockState.getMaterial().getColor().colorValue;
            }
        }

        BufferedImage pixelImage = new BufferedImage(zoomedSize, zoomedSize, BufferedImage.TYPE_INT_RGB);
        pixelImage.setRGB(0, 0, zoomedSize, zoomedSize, pixels, 0, zoomedSize);

        return TheMap.resizeImage(pixelImage, 320, 320);
    }

    private static int resolveCoorForChunk(int coor) {
        return coor < 0 ? ++coor : coor;
    }

    public static byte[] getChunkImageAsBytes(ServerWorld world, int x, int z, int zoom) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ImageIO.write(getChunkImage(world, x, z, zoom), "png", baos);
        baos.flush();
        byte[] byteArray = baos.toByteArray();
        baos.flush();
        return byteArray;
    }

    private static BufferedImage resizeImage(BufferedImage buf, int width, int height) {
        final BufferedImage bufImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        final Graphics2D g2 = bufImage.createGraphics();
        g2.drawImage(buf, 0, 0, width, height, null);
        g2.dispose();
        return bufImage;
    }

}


src/main/java/com/seirius/theeye/common/ChunkStorage.java
package com.seirius.theeye.common;

import net.minecraft.world.chunk.ChunkStatus;
import net.minecraft.world.chunk.IChunk;
import net.minecraft.world.server.ServerWorld;

import java.util.HashMap;

public class ChunkStorage {

    private final HashMap<String, IChunk> chunks = new HashMap<>();

    private final ServerWorld world;

    public ChunkStorage(ServerWorld world) {
        this.world = world;
    }

    public IChunk getChunk(int worldX, int worldZ) {
        int chunkX = worldX >> 4;
        int chunkZ = worldZ >> 4;
        String key = ChunkStorage.getKey(chunkX, chunkZ);
        IChunk chunk = chunks.get(key);
        if (chunk == null) {
            chunk = world.getChunk(chunkX, chunkZ, ChunkStatus.SURFACE);
            chunks.put(key, chunk);
        }
        return chunk;
    }

    public static String getKey(int chunkX, int chunkZ) {
        return String.format("%d:%d", chunkX, chunkZ);
    }

}


src/main/java/com/seirius/theeye/common/TestTheMap.java
package com.seirius.theeye.common;

import com.mojang.brigadier.CommandDispatcher;
import net.minecraft.block.BlockState;
import net.minecraft.command.CommandSource;
import net.minecraft.command.Commands;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.chunk.Chunk;
import net.minecraft.world.gen.Heightmap;
import net.minecraft.world.server.ServerWorld;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;

public class TestTheMap {

    public static int CHUNK_SIZE = 16;

    public static int MAX_ZOOM = 12;

    public static int transformZoom(int incommingZoom) {
        return MAX_ZOOM - incommingZoom + 1;
    }

    public static void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
            Commands.literal("testmap").requires((permission) -> permission.hasPermissionLevel(4))
                .then(Commands.literal("tiles").executes((command) -> {
                    try {
                        ServerWorld world = command.getSource().getWorld();
                        int chunkNumber = 16;
                        int size = chunkNumber * chunkNumber * CHUNK_SIZE;
                        int halfOffset = size / 2;
                        int[] pixels = new int[size * size];
                        for (int chunkX = -chunkNumber; chunkX < chunkNumber; chunkX++) {
                            for (int chunkZ = -chunkNumber; chunkZ < chunkNumber; chunkZ++) {
                                int minZ = chunkZ * CHUNK_SIZE;
                                int maxZ = minZ + CHUNK_SIZE;
                                int minX = chunkX * CHUNK_SIZE;
                                int maxX = minX + CHUNK_SIZE;
                                Chunk chunk = getChunk(world, minX, minZ);
                                for (int z = minZ; z < maxZ; z++) {
                                    for (int x = minX; x < maxX; x++) {
                                        int topY = chunk.getTopBlockY(Heightmap.Type.WORLD_SURFACE, x, z);
                                        BlockState blockState = chunk.getBlockState(new BlockPos(x, topY, z));
                                        pixels[(z + halfOffset) * size + (x + halfOffset)] = blockState.getMaterial().getColor().colorValue;
                                    }
                                }
                            }
                        }
                        BufferedImage pixelImage = new BufferedImage(size, size, BufferedImage.TYPE_INT_RGB);
                        pixelImage.setRGB(0, 0, size, size, pixels, 0, size);
                        File file = new File("test.png");
                        ImageIO.write(pixelImage, "png", file);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    return 1;
                }))
        );
    }

    private static Chunk getChunk(ServerWorld world, int x, int z) {
        return world.getChunkAt(new BlockPos(x, 0, z));
    }

    public static BufferedImage getChunkImage(ServerWorld world, int x, int z, int zoom) {
        int zoomedSize = CHUNK_SIZE * zoom;
        int[] pixels = new int[zoomedSize * zoomedSize];

        int worldStartZ = z * zoomedSize;
        int worldStartX = x * zoomedSize;

        int minZ = z * zoomedSize;
        int minX = x * zoomedSize;
        int currentChunkX = resolveCoorForChunk(minX) / CHUNK_SIZE;
        int currentChunkZ = resolveCoorForChunk(minZ) / CHUNK_SIZE;
        Chunk chunk = getChunk(world, minX, minZ);

        for (int imageZ = 0; imageZ < zoomedSize; imageZ++) {
            int worldZ = imageZ + worldStartZ;
            for (int imageX = 0; imageX < zoomedSize; imageX++) {
                int worldX = imageX + worldStartX;

                int newCurrentX = resolveCoorForChunk(worldX) / CHUNK_SIZE;
                int newCurrentZ = resolveCoorForChunk(worldZ) / CHUNK_SIZE;
                if (newCurrentX != currentChunkX || newCurrentZ != currentChunkZ) {
                    currentChunkX = newCurrentX;
                    currentChunkZ = newCurrentZ;
                    chunk = getChunk(world, worldX, worldZ);
                }
                int topY = chunk.getTopBlockY(Heightmap.Type.WORLD_SURFACE, worldX, worldZ);
                BlockState blockState = chunk.getBlockState(new BlockPos(worldX, topY, worldZ));
                pixels[imageZ * zoomedSize + imageX] = blockState.getMaterial().getColor().colorValue;
            }
        }

        BufferedImage pixelImage = new BufferedImage(zoomedSize, zoomedSize, BufferedImage.TYPE_INT_RGB);
        pixelImage.setRGB(0, 0, zoomedSize, zoomedSize, pixels, 0, zoomedSize);

        return TestTheMap.resizeImage(pixelImage, 320, 320);
    }

    private static int resolveCoorForChunk(int coor) {
        return coor < 0 ? ++coor : coor;
    }

    public static byte[] getChunkImageAsBytes(ServerWorld world, int x, int z, int zoom) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ImageIO.write(getChunkImage(world, x, z, zoom), "png", baos);
        baos.flush();
        byte[] byteArray = baos.toByteArray();
        baos.flush();
        return byteArray;
    }

    private static BufferedImage resizeImage(BufferedImage buf, int width, int height) {
        final BufferedImage bufImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        final Graphics2D g2 = bufImage.createGraphics();
        g2.drawImage(buf, 0, 0, width, height, null);
        g2.dispose();
        return bufImage;
    }

}
